File: KKProgressToolbar/KKProgressToolbar.m
//
// Copyright 2011-2012 Adar Porat (https://github.com/aporat)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import "KKProgressToolbar.h"

@interface KKProgressToolbar ()

@property(nonatomic, copy) KKProgressToolbarCompletionHandler showHandler;
@property(nonatomic, copy) KKProgressToolbarCompletionHandler hideHandler;

@end

@implementation KKProgressToolbar

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
        UIBarButtonItem *activityIndicatorItem = [[UIBarButtonItem alloc] initWithCustomView:self.activityIndicator];
        
        self.statusLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 2, 200, 20)];
        self.statusLabel.font = [UIFont boldSystemFontOfSize:14.0];
        self.statusLabel.backgroundColor = [UIColor clearColor];
        self.statusLabel.textColor = [UIColor whiteColor];
        self.statusLabel.shadowColor = [UIColor blackColor];
        self.statusLabel.shadowOffset = CGSizeMake(0, -1);
        self.statusLabel.textAlignment = UITextAlignmentCenter;
        
        self.progressBar = [[UIProgressView alloc] initWithFrame:CGRectMake(0, 25, 200, 10)];
        
        UIView *statusView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 200, 40)];
        [statusView addSubview:self.statusLabel];
        [statusView addSubview:self.progressBar];
        
        UIBarButtonItem *statusItem = [[UIBarButtonItem alloc] initWithCustomView:statusView];
        UIBarButtonItem *flexSpace = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
        self.stopButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemStop target:self action:@selector(didCancelButtonPressed:)];
        
        [self setBackgroundImage:nil
              forToolbarPosition:UIToolbarPositionAny
                      barMetrics:UIBarMetricsDefault];
        
        self.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleWidth;
        self.translucent = YES;
        self.barStyle = UIBarStyleBlackTranslucent;
        self.items = @[activityIndicatorItem, flexSpace, statusItem, flexSpace, self.stopButtonItem];
        
    }
    return self;
}

- (void)show:(BOOL)animated completion:(void (^)(BOOL finished))completion {
    self.showHandler = completion;
    
    self.stopButtonItem.enabled = YES;
    [self.activityIndicator startAnimating];

    [UIView animateWithDuration:0.4 animations:^{
        self.frame = CGRectMake(0, self.superview.bounds.size.height - 44, self.superview.bounds.size.width, 44);
    } completion:^(BOOL finished) {
        if (self.showHandler!=nil) {
            self.showHandler(YES);
        }
        self.showHandler = nil;
    }];
    
}

- (void)hide:(BOOL)animated completion:(void (^)(BOOL finished))completion {
    self.hideHandler = completion;
    
    self.stopButtonItem.enabled = NO;
    [self.activityIndicator stopAnimating];
    
    [UIView animateWithDuration:0.4 delay:1.0 options:UIViewAnimationOptionLayoutSubviews animations:^{
        self.frame = CGRectMake(0, self.superview.bounds.size.height, self.superview.bounds.size.width, 44);
    } completion:^(BOOL finished) {
        if (self.hideHandler!=nil) {
            self.hideHandler(YES);
        }
        self.hideHandler = nil;
    }];
    
}

- (void)didCancelButtonPressed:(id*)sender {
    if ([_actionDelegate respondsToSelector:@selector(didCancelButtonPressed:)]) {
        [_actionDelegate performSelector:@selector(didCancelButtonPressed:) withObject:self];
    }
}



@end

File: KKProgressToolbar/KKProgressToolbar.swift
import UIKit
import GTProgressBar
import SnapKit

public protocol KKProgressToolbarDelegate: class {
    func didCancelButtonPressed(_ toolbar: KKProgressToolbar)
}

public final class KKProgressToolbar: UIView {
    
    fileprivate var didSetupConstraints = false
    
    // MARK: - Data
    weak public var actionDelegate: KKProgressToolbarDelegate?
    public var isShown = false
    
    public var text: String? {
        didSet {
            titleLabel.text = text
        }
    }
    
    // MARK: - UI Elements
    lazy fileprivate var stopButton: UIButton = {
        let button = UIButton(type: .custom)
        button.backgroundColor = .clear
        button.setImage(UIImage(named: "toolbar-stop"), for: .normal)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.addTarget(self, action: #selector(didCancelButtonPressed(_:)), for: .touchUpInside)
        return button
    }()
    
    // MARK: - UI Elements
    lazy fileprivate var mainBackgroundView: UIView = {
        let view = UIView()
        view.backgroundColor = .black
        view.alpha = 0.75
        
        return view
    }()
    
    lazy fileprivate var titleLabel: UILabel = {
        let view = UILabel()
        view.font = UIFont.systemFont(ofSize: 13)
        view.textColor = .white
        view.textAlignment = .center
        view.backgroundColor = .clear
        
        return view
    }()
    
    lazy public var progressBar: GTProgressBar = {
        let view = GTProgressBar()
        view.displayLabel = true
        view.font = UIFont.systemFont(ofSize: 13)
        view.labelTextColor = .white
        view.progressLabelInsetRight = 15
        view.barFillInset = 1
        
        return view
    }()
    
    // MARK: - UIView
    required override init(frame: CGRect) {
        super.init(frame: frame)
        setupSubviews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override public func awakeFromNib() {
        super.awakeFromNib()
        setupSubviews()
    }
    
    override public func updateConstraints() {
        super.updateConstraints()
        
        if !didSetupConstraints {
            
            mainBackgroundView.snp.makeConstraints { make in
                make.edges.equalTo(self)
            }
            
            stopButton.snp.makeConstraints { make in
                make.centerY.equalTo(self)
                make.size.equalTo(25)
                make.trailing.equalTo(self).offset(-14)
            }
            
            titleLabel.snp.makeConstraints { make in
                make.top.equalTo(self).offset(5)
                make.height.equalTo(16)
                make.centerX.equalTo(self)
            }
            
            progressBar.snp.makeConstraints { make in
                make.top.equalTo(titleLabel.snp.bottom).offset(2)
                make.height.equalTo(12)
                make.leading.equalTo(self).offset(14)
                make.trailing.equalTo(stopButton.snp.leading).offset(-10)
            }
            
            didSetupConstraints = true
        }
    }
    
    fileprivate func setupSubviews() {
        addSubview(mainBackgroundView)
        addSubview(stopButton)
        addSubview(titleLabel)
        addSubview(progressBar)
        
        setNeedsUpdateConstraints()
    }
    
    public func show(_ animated: Bool, completion: (() -> Void)? = nil) {
        if !isShown {
            isShown = true
            
            stopButton.isEnabled = true
            
            if let superview = self.superview {
                self.frame = CGRect(x: 0, y: superview.bounds.size.height, width: superview.bounds.size.width, height: 55)
            }
            
            if animated {
                UIView.animate(withDuration: 0.4, delay: 0.0, options: [], animations: {
                    if let superview = self.superview {
                        self.frame = CGRect(x: 0, y: superview.bounds.size.height - 55, width: superview.bounds.size.width, height: 55)
                    }
                }) { _ in
                    self.isHidden = false
                    completion?()
                }
            } else {
                if let superview = self.superview {
                    self.frame = CGRect(x: 0, y: superview.bounds.size.height - 55, width: superview.bounds.size.width, height: 55)
                    self.isHidden = false
                    completion?()
                }
            }
        } else {
            completion?()
        }
    }
    
    public func hide(_ animated: Bool, completion: (() -> Void)? = nil) {
        if isShown {
            isShown = false
            
            stopButton.isEnabled = false
            
            if animated {
                UIView.animate(withDuration: 0.4, delay: 1.0, options: [], animations: {
                    if let superview = self.superview {
                        self.frame = CGRect(x: 0, y: superview.bounds.size.height, width: superview.bounds.size.width, height: 55)
                    }
                }) { _ in
                    self.isHidden = true
                    completion?()
                }
            } else {
                if let superview = self.superview {
                    self.frame = CGRect(x: 0, y: superview.bounds.size.height, width: superview.bounds.size.width, height: 55)
                    self.isHidden = true
                    completion?()
                }
            }
        } else {
            completion?()
        }
    }
    
    @objc public func didCancelButtonPressed(_ sender: Any?) {
        actionDelegate?.didCancelButtonPressed(self)
    }
    
}
